src/main/java/hudson/plugins/git/GitSCM.java

```diff
@@ -583,22 +583,28 @@ private PollingResult compareRemoteRevisionWithImpl(Job<?, ?> project, Launcher
                     for (BranchSpec branchSpec : getBranches()) {
                         for (Entry<String, ObjectId> entry : heads.entrySet()) {
                             final String head = entry.getKey();
-                            String name;
                             // head is "refs/(heads|tags)/branchName
-                            if (head.startsWith("refs/heads/")) name = remote + "/" + head.substring(11);
-                            else if (head.startsWith("refs/tags/")) name = remote + "/" + head.substring(10);
-                            else name = remote + "/" + head;
 
-                            if (!branchSpec.matches(name, environment)) continue;
+                            // first, check the a canonical git reference is configured
+                            if (!branchSpec.matches(head, environment)) {
+
+                                // convert head `refs/*` into shortcut notation `remote/branch`
+                                String name = head;
+                                if (head.startsWith("refs/heads/")) name = remote + "/" + head.substring(11);
+                                else if (head.startsWith("refs/tags/")) name = remote + "/" + head.substring(10);
```

Comment:
Again hardcode, this wouldn't work with `ref/notes` or any other git based solution. `refs/heads` is just a DEFAULT for git branches.

I can propose add new UI specifier as "branch specifier", but some 'git refspec specifier' that will work right as git documentation, move matching to this class.